File: typescript/src/utils/serializer.ts
import { TerraformDocument } from '../types/blocks';
import { TerraformExport } from '../types/artifacts';
import { createExport } from './graphBuilder';
import { toYaml } from './yaml';

export interface SerializeOptions {
    pruneEmpty?: boolean;
}

export function toJson(document: TerraformDocument | unknown, options?: SerializeOptions): string {
    const value = shouldPrune(options) && isTerraformDocument(document) ? pruneDocument(document) : document;
    return JSON.stringify(value, null, 2);
}

export function toJsonExport(document: TerraformDocument, options?: SerializeOptions): string {
    return JSON.stringify(toExport(document, options), null, 2);
}

export function toExport(document: TerraformDocument, options?: SerializeOptions): TerraformExport {
    const exportPayload = createExport(document);
    const prunedDocument = shouldPrune(options) ? pruneDocument(document) : document;
    return { ...exportPayload, document: prunedDocument };
}

export function toYamlDocument(document: TerraformDocument | unknown, options?: SerializeOptions): string {
    const value = shouldPrune(options) && isTerraformDocument(document) ? pruneDocument(document) : document;
    return toYaml(value);
}

function shouldPrune(options?: SerializeOptions): boolean {
    return options?.pruneEmpty !== false;
}

function pruneDocument(document: TerraformDocument): Partial<TerraformDocument> {
    return (pruneValue(document) as Partial<TerraformDocument>) ?? {};
}

function pruneValue(value: unknown): unknown {
    if (value === null || value === undefined) {
        return undefined;
    }

    if (Array.isArray(value)) {
        const items = value
            .map((item) => pruneValue(item))
            .filter((item) => item !== undefined) as unknown[];
        return items.length > 0 ? items : undefined;
    }

    if (typeof value === 'object') {
        const entries = Object.entries(value as Record<string, unknown>);
        const pruned: Record<string, unknown> = {};

        for (const [key, val] of entries) {
            const next = pruneValue(val);
            if (next === undefined) {
                continue;
            }

            if (Array.isArray(next) && next.length === 0) {
                continue;
            }
            if (typeof next === 'object' && next !== null && Object.keys(next as Record<string, unknown>).length === 0) {
                continue;
            }

            pruned[key] = next;
        }

        return Object.keys(pruned).length > 0 ? pruned : undefined;
    }

    return value;
}

function isTerraformDocument(doc: unknown): doc is TerraformDocument {
    return Boolean(doc && typeof doc === 'object' && 'resource' in (doc as Record<string, unknown>));
}

File: python/src/parse_hcl/serializer.py
from __future__ import annotations

import json
from typing import Any, Dict

from .graph import create_export
from .types import TerraformDocument, TerraformExport
from .yaml_serializer import to_yaml


def to_json(document: Any, prune_empty: bool = True) -> str:
    value = _prune_document(document) if prune_empty and _is_terraform_document(document) else document
    return json.dumps(value, indent=2)


def to_json_export(document: TerraformDocument, prune_empty: bool = True) -> str:
    return json.dumps(to_export(document, prune_empty=prune_empty), indent=2)


def to_export(document: TerraformDocument, prune_empty: bool = True) -> TerraformExport:
    export_payload = create_export(document)
    pruned_document = _prune_document(document) if prune_empty else document
    export_payload["document"] = pruned_document
    return export_payload


def to_yaml_document(document: Any, prune_empty: bool = True) -> str:
    value = _prune_document(document) if prune_empty and _is_terraform_document(document) else document
    return to_yaml(value)


def _prune_document(document: TerraformDocument) -> Dict[str, Any]:
    return _prune_value(document) or {}


def _prune_value(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, list):
        items = [item for item in (_prune_value(item) for item in value) if item not in (None, [], {}, ())]
        return items or None
    if isinstance(value, dict):
        pruned: Dict[str, Any] = {}
        for key, val in value.items():
            next_val = _prune_value(val)
            if next_val is None or next_val == [] or next_val == {}:
                continue
            pruned[key] = next_val
        return pruned or None
    return value


def _is_terraform_document(doc: Any) -> bool:
    return isinstance(doc, dict) and "resource" in doc
